### ISSUE TITLE
Ideas on design

### ISSUE THREAD

**Issue by evetion** [JoshuaBillson/RemoteSensingToolbox.jl#34]
Hi there 👋🏻 

I have happily used this package in the past week (see https://github.com/evetion/OGH2023/blob/main/hackathons/landcover.ipynb). It's just what I needed for the landcover classification hackathon there, using MLJ (see https://www.kaggle.com/competitions/ogh2023). Thanks 👍🏻 

Not sure if you need it, but some thoughts, also looking at your planned features.

There's the following code
```julia
landsat = read_bands(Landsat8, "LC08_L2SP_043024_20200802_20200914_02_T1/")
ndvi(landsat, Landsat8)
```

Ideally if one reads a landsat bandset, the rasterstack would "know" it's a Landsat8 stack. Maybe a wrapper type would help here. With a type, one can dispatch (normally on the instance of the type, so `Landsat8()`), so one can just do `ndvi(landsat)`. Implementations would then follow `ndvi(::Landsat8, rasterstack)` (the type to dispatch on is often first).

In terms of ML, I don't see a clear path yet, it's mostly preprocessing I guess? It's DataFrame(rasterstack), dropping the dimension columns and making sure that there are no NaNs/missing/Infs. Something that should be done with the dimension columns is making sure we can do spatial cross-validation (something that exists in GeoStats.jl, but not in the MLJ ecosystem I believe).

**Comment by JoshuaBillson** [comment]
Hey, thanks for the feedback! I'm glad to hear others are finding it useful!

Regarding the bandset types, I initially tried what you proposed (wrapping `Raster` and `RasterStack` objects inside a struct), but I realized a couple months in that there would be problems with long-term maintainability. Basically, to make a `Landsat8` or any other bandset compatible with external methods like `mask`, `crop`, or `resample`, we'd need to write a specialized method for each that unwraps and wraps the contained raster. 

While this is doable for a few packages, it becomes problematic when trying to support all packages that expect a `Raster` or `RasterStack` as input. Forcing users to pass the sensor type explicitly was a tradeoff I made for the sake of maintenance and extensibility. However, I do agree it would be ideal for a raster to know what type of sensor it belongs to so users don't need to repeatedly pass it in every time. One possibility I've toyed with is attaching this information to the raster's metadata field when it's first loaded. However, I'm not sure how robust the support for this is in [Rasters.jl](https://github.com/rafaqz/Rasters.jl) since it doesn't seem to be currently documented.

Regarding ML frameworks, I agree that `GeoStats.jl` is the way forward. The main problem is that the existing `Tables.jl` interface provided by `Rasters.jl` has the following issues:
1. Doesn't treat `Band` dimension like layers when working with multiband images.
2. Doesn't automatically replace or drop missing values.
3. Poorly optimized for working with a large number of layers, which is the case for hyperspectral imagery.
4. No obvious way to reconstruct a `Raster` or `RasterStack` from a table with a `:geometry` column containing `Meshes.jl` geometies.

To address this, I've been working on an internal table type called `RasterTable`. This should demonstrate how it's intended to work:
```julia
julia> t = RasterTable(landsat) |> RemoteSensingToolbox.dropmissing
RasterTable with 40483825 rows, 9 columns, and schema:
 :X   Float64
 :Y   Float64
 :B1  UInt16
 :B2  UInt16
 :B3  UInt16
 :B4  UInt16
 :B5  UInt16
 :B6  UInt16
 :B7  UInt16

julia> geotable = georef(t, (:X, :Y))
40483825 PointSet{2,Float64}
  variables (rank 0)
    └─B1 (UInt16)
    └─B2 (UInt16)
    └─B3 (UInt16)
    └─B4 (UInt16)
    └─B5 (UInt16)
    └─B6 (UInt16)
    └─B7 (UInt16)
```

As you can see, we can easily convert a `Raster` or `RasterStack` to a table compatible with `GeoStats`. The next task will be to somehow convert a table of predictions back into a `Raster`. Once that's done, I think we can provide a simple interface to allow users to predict land cover classes with any `GeoStats` solver and `MLJ` model they desire.

**Comment by evetion** [comment]
@rafaqz

**Comment by rafaqz** [comment]
Thanks for the writeup this is all good to know!

We should aim as much as possible to fix these problems in DimensionalData.jl and Rasters.jl where they help everyone, rather than working around them here.

### Metadata

You could totally put the landsat details in `metadata` thats really why its there. Its how netcdf and gdal metadata propagates between read and write, so its very much supported in Rasters. It may need some tweaking for your use case but at least try it.

(When there are missing docs and features, your first guess should be we never had time to write them. Rasters is one of very many packages I maintain in spare time, and I have to work on all the underying deps like DiskArrays, DimensionalData, GeoInterface, ArchGDAL etc for things to work at all)

### Other issues

1. Not exactly sure what you mean here, but this is maybe doable already, load the tif as a `RasterStack`, there is a `layerfrom=Band` keyword by default which will turn bands into layers of the stack

2. Auto replacing missing is the easiest of these to implement in Rasters, by just running `replace_missing` in the constructors, with a keyword to turn it off. Its pretty much next task on my list. Dropping missing from tables isn't easy generically because missing may not be in the same place in every layer. If you have an algorithm to detect that it could also go in DD.

3. The slow-many-layers problem is a known issue. We could fix it in DD by addding a Vector based `MDimStack` (mutable) rather than tuple based `DimStack`. But thats a _lot_ of work, so hasnt been done yet.

4. Again I just never got around to it in DD, would love to get a `DimStack` back from a table. AxisKeys.jl has some similar code for this.

**Comment by evetion** [comment]
> Regarding ML frameworks, I agree that `GeoStats.jl` is the way forward. The main problem is that the existing `Tables.jl` interface provided by `Rasters.jl` has the following issues:
> 
>     4. No obvious way to reconstruct a `Raster` or `RasterStack` from a table with a `:geometry` column containing `Meshes.jl` geometies.

My remark about GeoStats.jl was mostly about *spatial* cross-validation, for which we need to keep a geometry around, although a MLJ model probably doesn't need to know about it. As long as we implement Tables and GeoInterface (zipped X, Y columns) interfaces, most of the use-cases should work out of the box.

Regarding the way back to Rasters (only useful for predictions right, not for training), we should either store a logical index, or make sure that we keep all rows (skipping over them if they contain NaNs?). In the notebook I linked I just reshaped the prediction vector back to the raster.

**Comment by rafaqz** [comment]
There is a way to reshape after dropping missings - just look up the x/y values from the table columns. Of course its much slower than a reshape but not too bad.

DimensionalData could also zip X and Y values to a single NamedTuple column to be GeoInterface compatible (the implementation preadates the current GeoInterface) rather than the separate columns it does now - we could add a keyword for `DimTable` that allowed that and set it by default in Rasters.

**Comment by JoshuaBillson** [comment]
> ### Other issues
> 1. Not exactly sure what you mean here, but this is maybe doable already, load the tif as a `RasterStack`, there is a `layerfrom=Band` keyword by default which will turn bands into layers of the stack
> 2. Auto replacing missing is the easiest of these to implement in Rasters, by just running `replace_missing` in the constructors, with a keyword to turn it off. Its pretty much next task on my list. Dropping missing from tables isn't easy generically because missing may not be in the same place in every layer. If you have an algorithm to detect that it could also go in DD.
> 3. The slow-many-layers problem is a known issue. We could fix it in DD by addding a Vector based `MDimStack` (mutable) rather than tuple based `DimStack`. But thats a _lot_ of work, so hasnt been done yet.
> 4. Again I just never got around to it in DD, would love to get a `DimStack` back from a table. AxisKeys.jl has some similar code for this.

1. Turning everything into a `RasterStack` would be one solution and is what I initially tried. However, `RasterStack` is poorly optimized for imagery with hundreds of bands, as is the case for hyperspectral sensors. I think this is because `RasterStack`  exposes each layer as a type parameter, which Julia struggles to compile. We need something like `DataFrame`, which drops the type specification for each column to enable operations on wide tables.
2. My algorithm for dropping missing values involves finding the set intersection of non-missing rows in each column. Using `BitSet` for this purpose seems to yield a substantial performance increase. Regarding replacing missing values directly in the rasters, would this incur a significant performance penalty? I would imagine Julia would have a harder time operating on arrays of `Union{UInt16,Missing}` than `UInt16`.
3. I recall having this conversation on [Rasters.jl](https://github.com/rafaqz/Rasters.jl/issues/464). I looked into implementing it myself, but to be honest, I didn't know where to start. Adding support for multiband rasters seemed like the most prudent solution at the time.
4. I should note that it's already possible to reconstruct both a `Raster` or `RasterStack` from a table. However, it only works if no rows have been dropped and their order is unchanged.

At this point in time, having a robust and efficient method for converting rasters to tables and back again would probably be one of the most valuable features to implement. This would open up the possibility of using rasters in the wider ML ecosystem as well as computing statistics and transformations. Currently, these tasks are somewhat tedious.

**Comment by rafaqz** [comment]
1. Yes totally, we need the vector based version for that to work with RasterStack. Another option is using a keyword to make colums out of a dimension of a Raster in `DimTable`. Thats not actually hard to do, maybe even only a few lines of code.
2. Yes. This is the reason to not automatically replace, its slow.
3. Yeah, theyre complicated internals
4. You can also to it with changed order using the `At` selector and `setindex!` in a loop over the rows. But its slow.

Lets do 1 first then with a keyword to `DimTable`, thats any easy win

**Comment by JoshuaBillson** [comment]
One problem with `DimTable` is that it exposes the columns in the type parameters. This makes it slow for tables with several hundred columns, which makes it impractical for hyperspectral images. It's for this reason that `RasterTable` uses a vectorized representation:

```julia
mutable struct RasterTable <: Tables.AbstractColumns
    layers::Vector{Symbol}
    cols::Vector{Vector}
end

Tables.schema(table::RasterTable) = Tables.Schema{nothing,nothing}(layers(table), [eltype(c) for c in cols(table)])
```

This is the result of trying to convert a DESIS image with 235 bands into a `DimTable`:
```julia
julia> @time Rasters.DimTable(desis);
 17.052105 seconds (699.42 k allocations: 620.133 MiB, 14.31% gc time, 85.05% compilation time)
```

**Comment by rafaqz** [comment]
Right, guess we will have to do the same for DimTable as for DimStack. But its much less code for DimTable. Or I guess MDimTable or something.

But yeah, esentially your RasterTable work may as well be in DD so YAXArrays.jl, Arviz.jl and everyone else gets to use it too.

**Comment by JoshuaBillson** [comment]
Yes, I think we need a fairly comprehensive overhaul of `DimTable`. I've opened an [issue](https://github.com/rafaqz/DimensionalData.jl/issues/535) on `DimensionalData` where we can discuss the specifics. Ideally, we should be able to create a general solution for feeding rasters into ML workflows, which would remove the need to implement it in this package.
